### PR TITLE
WN-TF-Filter-List

### LINKED ISSUE #6: 6. As a user, I want to filter my shopping list to make it easier to locate an item in the list.
## Summary
Allowing users to search through their list will help them cut down on clutter and focus on items they’re looking for.

## Acceptance criteria
- [ ] A form is added to the top of the `List` view, above the shopping list
- [ ] The form includes the following elements
	- A text field (with semantic `<label>`!) which narrows down the list as the user types
	- When there’s text in the field, some kind of button (e.g., an X) to clear the field. When the field is cleared, the list is reset to its unfiltered state.

## Notes
- You will know your `<label>` is correctly implemented if you click on the label and keyboard focus moves to the related input. Refer to [the MDN docs on `<label>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label) for more info.
- The filter text should match *any* part of the item name (e.g., it should _not_ only match from the start of the string)
- [This wireframe](https://github.com/the-collab-lab/smart-shopping-list/blob/main/_resources/wireframes/list-populated.png) shows the filter field, but not it being used

### PR DESCRIPTION
**Pull request by wheninseattle**
## Description

This code added a form for users to filter list items. The form includes a reset button, which appears when a search query is present. As the user types, the list will render all items that include any part of the search query. If no items match, the component conditionally renders a message to provide users with the feedback that no items match their search.

## Approach

- We used the useState() hook, so that the search term would be updated dynamically.
- We opted to pull out the filter function into a helper function, so it could double as a condition for the render.

## Lessons Learned
- We experienced an issue of the incorrect jsx rendering momentarily upon the list component's initial load. This was happening because the condition being evaluated was the length of the array filteredListItems. Upon the initial load, the component renders prior to the data being ready for use. We addressed this by using the || operator to add a test of the search query being empty.
- We initially used onReset() in our form, however, when the user pressed 'enter' it triggered the form reset, which resulted in the full list being reloaded. Users would likely expect 'enter' to complete their search, so this could be confusing. To address this, we changed to onClick and included an onSubmit handler to prevent default form submission behavior. 

## Related Issue

closes #6 

## Acceptance Criteria

- [x] A form is added to the top of the `List` view, above the shopping list
- [x] The form includes the following elements
	- A text field (with semantic `<label>`!) which narrows down the list as the user types
	- When there’s text in the field, some kind of button (e.g., an X) to clear the field. When the field is cleared, the list is reset to its unfiltered state.
	- 
## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

### Before

image

### After
image

## Testing Steps / QA Criteria

1. git checkout WN-TF-Filter-List
2. npm start
3. Navigate to http://localhost:3000/list
4. Click the input label to test it
5. Type a search query into the provided input field
6. Enjoy viewing your curated list!

### CODE REVIEW

**Review by a-cole55 — APPROVED** [review]
@tmflower @wheninseattle great job! Your code is very clear and precise. Easy to read and understand. I'm really enjoying the filter feature.

File: src/views/List.jsx [hunk 1]
```diff
@@ -1,14 +1,52 @@
...
-			</ul>
+			<form onSubmit={handleSubmit}>
+				<label htmlFor="searchTerm">Filter list items</label>
+				<input
+					type="text"
+					value={searchTerm}
+					id="searchTerm"
+					placeholder="Start typing here..."
+					onChange={onSearch}
+				></input>
```
Comment (mariahschock) [inline src/views/List.jsx]: The smallest of critiques... but you could close the input on line 34 with just the  />
I think this would make it look just a little cleaner right here :)

**Review by mariahschock — APPROVED** [review]
The code looks clean and is functioning appropriately. Easy to read. I also liked how you made it so the user doesn't have to be case-sensitive when searching for an item.

File: src/views/List.jsx (hunk 1 above)
Comment (wheninseattle) [reply src/views/List.jsx]: Oh, great call @mariahschock. I went ahead and made that change. All 'nits' welcome!

File: src/views/List.jsx [hunk 2]
```diff
@@ -1,14 +1,52 @@
...
+					value={searchTerm}
+					id="searchTerm"
+					placeholder="Start typing here..."
+					onChange={onSearch}
+				/>
+				{searchTerm === '' ? null : (
+					<button onClick={onSearchReset} type="reset">
+						Reset
+					</button>
+				)}
```
Comment (veeteeran) [inline src/views/List.jsx:39]: optional we could shorten this to be:
```javascript
{!!searchterm &&
<button onClick={onSearchReset} type="reset">
    Reset
</button>}
```

File: src/views/List.jsx [hunk 3]
```diff
@@ -1,14 +1,52 @@
...
+			</form>
+			{filteredData.length > 0 || !searchTerm ? (
```
Comment (veeteeran) [inline src/views/List.jsx]: optional same as above
```javascript
!!filteredData.length || !searchTerm ? (
...
)
```

File: src/views/List.jsx (hunk 2 above)
Comment (veeteeran) [inline src/views/List.jsx]: Very optional. What you have written is great!! We could shorten this to be:
```javascript
{!!searchterm &&
<button onClick={onSearchReset} type="reset">
    Reset
</button>
}
```

The double !! can be a little confusing so try this only if you feel up to it. [Stackoverflow](https://stackoverflow.com/questions/29312123/how-does-the-double-exclamation-work-in-javascript)

File: src/views/List.jsx [hunk 4]
```diff
@@ -1,14 +1,52 @@
...
+				<label htmlFor="searchTerm">Filter list items</label>
+				<input
+					type="text"
+					value={searchTerm}
+					id="searchTerm"
+					placeholder="Start typing here..."
+					onChange={onSearch}
+				/>
+				{searchTerm === '' ? null : (
+					<button onClick={onSearchReset} type="reset">
```
Comment (veeteeran) [inline src/views/List.jsx:36]: Great use of type="reset" 😸

**Review by veeteeran — APPROVED** [review]
Well done Weston and Tina!!! This code is so clean! You both are using ternaries efficiently for the conditional renders.  Having `filteredList` variable helps so much with readability later in the code. Only some very optional suggestions to use the double bang. I am super impressed!!

File: src/views/List.jsx (hunk 3 above)
Comment (wheninseattle) [reply src/views/List.jsx]: Thanks for suggesting this @veeteeran!! I didn't know about this, and it was interesting to read about.

File: src/views/List.jsx (hunk 2 above)
Comment (wheninseattle) [reply src/views/List.jsx]: Thanks, I really like how clean this is.
